### DISCUSSION TITLE
About warming up capsules on backend

### DISCUSSION THREAD

**Discussion by busslina** [GregoryConrad/rearch-dart#125] [General]
Can you suggest a technique to warming up capsules in a backend app?

**Comment by GregoryConrad** [comment]
Something like the following should work I'd think, but hasn't been tested.

```dart
Future<void> main() async {
  final container = CapsuleContainer()..read(aWarmUpCapsule)..read(bWarmUpCapsule);
  await (use(aAsyncCapsule), use(bAsyncCapsule)).wait;
  // do whatever with aCapsule and bCapsule
}

Future<int> aAsyncCapsule(CapsuleHandle use) async => throw 'impl hidden';
Future<bool> bAsyncCapsule(CapsuleHandle use) async => throw 'impl hidden';
AsyncValue<int> aWarmUpCapsule(CapsuleHandle use) => use.future(use(aAsyncCapsule));
AsyncValue<bool> bWarmUpCapsule(CapsuleHandle use) => use.future(use(bAsyncCapsule));
int aCapsule(CapsuleHandle use) => (use(aWarmUpCapsule) as AsyncData).data;
bool bCapsule(CapsuleHandle use) => (use(bWarmUpCapsule) as AsyncData).data;
```

**Reply by GregoryConrad** [reply]
If that doesn't work, you can manually create a future and use container.listen on the `xyzWarmUpCapsule`s to set the future as completed once they are an AsyncData.

**Comment by busslina** [comment]
Seems a cool way !

**Reply by busslina** [reply]
Worked

Capsules:
```dart
Future<String> hostnameAsyncCapsule(CapsuleHandle use) =>
    use(defaultShellCapsule).hostname();

AsyncValue<String> hostnameWarmUpCapsule(CapsuleHandle use) =>
    use.future(use(hostnameAsyncCapsule));

String hostnameCapsule(CapsuleHandle use) =>
    use(hostnameWarmUpCapsule).data.asNullable()!;
```

Warm up:
```dart
cc.read(hostnameWarmUpCapsule);
await Future.wait([cc.read(hostnameAsyncCapsule)]);
```

**Reply by GregoryConrad** [reply]
Great! Glad to hear it.

**Reply by busslina** [reply]
Even more refined:

Typedef + extension:
```dart
typedef WarmUpCapsulePair<T> = (
  Capsule<Future<T>>,
  Capsule<AsyncValue<T>>,
);

extension WarmUpCapsulePairExtension<T> on WarmUpCapsulePair<T> {
  Future<T> warmUp() {
    cc.read($2);
    return cc.read($1);
  }
}
```

Capsules:
```dart
Future<String> hostnameFutureCapsule(CapsuleHandle use) =>
    use(defaultShellCapsule).hostname();

AsyncValue<String> hostnameAsyncCapsule(CapsuleHandle use) =>
    use.future(use(hostnameFutureCapsule));

String hostnameCapsule(CapsuleHandle use) =>
    use(hostnameAsyncCapsule).data.asNullable()!;
```

Warm up:
```dart
await Future.wait([
        (hostnameFutureCapsule, hostnameAsyncCapsule).warmUp(),
      ]);
```

**Reply by opsb** [reply]
It seems a bit awkward that you need to wait on both the warm up and async capsules. Is there an approach where we could mirror the toWarmUpCapsule and only specify the warm up capsules?

**Reply by GregoryConrad** [reply]
@opsb Sure, you could:
1. Manually create a future (can't remember the Dart API for this since I needed it once, probably some 6 years ago now)
2. `container.listen` to just the warm up capsules (the ones that return an `AsyncValue`)
3. Complete the future when all of the warm up capsules are `AsyncData`. Error the future when any of the capsules are `AsyncError`

I just never bothered to write that out. If you wanted to add that in a PR along with a test case or two I'd be happy to incorporate it into the framework (the above 3 steps are the proper way to actually warm something up; what I gave before is the "quick and dirty" way to do it in less work).

Could likely be an `extension WarmUpCapsuleContainer on CapsuleContainer { Future<void> warmUp<T>(Iterable<Capsule<T>> capsules) => throw 'todo'; }`

**Reply by busslina** [reply]
> can't remember the Dart API

https://api.flutter.dev/flutter/dart-async/Completer-class.html
